imod/mf6/rch.py

```diff
@@ -152,3 +161,80 @@ def _validate(self, schemata, **kwargs):
 
     def get_regrid_methods(self) -> Optional[dict[str, Tuple[RegridderType, str]]]:
         return self._regrid_method
+
+    @classmethod
+    def from_imod5_data(
+        cls,
+        imod5_data: dict[str, dict[str, GridDataArray]],
+        target_grid: GridDataArray,
+        regridder_types: Optional[dict[str, tuple[RegridderType, str]]] = None,
+    ) -> "Recharge":
+        """
+        Construct an rch-package from iMOD5 data, loaded with the
+        :func:`imod.formats.prj.open_projectfile_data` function.
+
+        .. note::
+
+            The method expects the iMOD5 model to be fully 3D, not quasi-3D.
+
+        Parameters
+        ----------
+        imod5_data: dict
+            Dictionary with iMOD5 data. This can be constructed from the
+            :func:`imod.formats.prj.open_projectfile_data` method.
+        target_grid: GridDataArray
+            The grid that should be used for the new package. Does not
+            need to be identical to one of the input grids.
+        regridder_types: dict, optional
+            Optional dictionary with regridder types for a specific variable.
+            Use this to override default regridding methods.
+
+        Returns
+        -------
+        Modflow 6 rch package.
+
+        """
+
+        data = {
+            "rate": imod5_data["rch"]["rate"],
+        }
+        new_package_data = {}
+
+        # first regrid the inputs to the target grid.
+        regridder_settings = deepcopy(cls._regrid_method)
+        if regridder_types is not None:
+            regridder_settings.update(regridder_types)
+
+        regrid_context = RegridderWeightsCache()
+
+        new_package_data = _regrid_package_data(
+            data, target_grid, regridder_settings, regrid_context, {}
+        )
+
+        # if rate has only layer 0, then it is planar.
+        if is_planar_grid(data["rate"]):
+            # create an array indicating in which cells rch is active
+            is_rch_cell = allocate_rch_cells(
+                ALLOCATION_OPTION.at_first_active,
+                target_grid,
+                new_package_data["rate"].isel(layer=0),
+            )
+            rch_rate = zeros_like(target_grid)
+
+            # assign the rch-rate to all cells in a column
+            if is_transient_data_grid(data["rate"]):
+                rch_rate = deepcopy(
+                    rch_rate.expand_dims({"time": data["rate"]["time"]})
+                )
+                for time in data["rate"]["time"].values:
+                    rch_rate.loc[time, :, :, :] = (
+                        data["rate"].sel(time=time).drop_vars("layer")
+                    )
+            else:
+                rch_rate.loc[:, :, :] = data["rate"].drop_vars("layer")
+
+            # remove rch from cells where it is not allocated
+            rch_rate = xr.where(is_rch_cell, rch_rate, np.nan)
```

Comment:
I think we can better utilize xarray's broadcasting here:
- We do not have to load each timestep into memory and assign manually.
- Code can be reduced
- You are assigning the old ``data["rate"]`` instead of the regridded ``new_package_data["rate"]`` to ``rch_rate``

I might be naive, but I think we can reduce these data transformations to:

```suggestion

            # remove rch from cells where it is not allocated and broadcast over layers.
            rch_rate = new_package_data["rate"].where(is_rch_cell)
            rch_rate = enforce_dim_order(rch_rate)
```

Reply:
nice, much better